pageserver/src/tenant/timeline.rs

```diff
@@ -4235,87 +4237,147 @@ impl Timeline {
+                                        } else {
+                                            return Err(
+                                                CreateImageLayersError::PageReconstructError(err),
+                                            );
+                                        }
                                     }
-                                }
-                            };
+                                };
 
-                            // Write all the keys we just read into our new image layer.
-                            image_layer_writer.put_image(img_key, img, ctx).await?;
-                            wrote_keys = true;
+                                // Write all the keys we just read into our new image layer.
+                                image_layer_writer.put_image(img_key, img, ctx).await?;
+                                wrote_keys = true;
+                            }
                         }
                     }
                 }
-            }
 
-            if wrote_keys {
-                // Normal path: we have written some data into the new image layer for this
-                // partition, so flush it to disk.
-                start = img_range.end;
-                let image_layer = image_layer_writer.finish(self, ctx).await?;
-                image_layers.push(image_layer);
+                if wrote_keys {
+                    // Normal path: we have written some data into the new image layer for this
+                    // partition, so flush it to disk.
+                    start = img_range.end;
+                    let image_layer = image_layer_writer.finish(self, ctx).await?;
+                    image_layers.push(image_layer);
+                } else {
+                    // Special case: the image layer may be empty if this is a sharded tenant and the
+                    // partition does not cover any keys owned by this shard.  In this case, to ensure
+                    // we don't leave gaps between image layers, leave `start` where it is, so that the next
+                    // layer we write will cover the key range that we just scanned.
+                    tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+                }
             } else {
-                // Special case: the image layer may be empty if this is a sharded tenant and the
-                // partition does not cover any keys owned by this shard.  In this case, to ensure
-                // we don't leave gaps between image layers, leave `start` where it is, so that the next
-                // layer we write will cover the key range that we just scanned.
-                tracing::debug!("no data in range {}-{}", img_range.start, img_range.end);
+                // Metadata keys image layer creation.
+                let total_kb_reads_begin = ctx
+                    .vectored_access_delta_file_size_kb
+                    .load(AtomicOrdering::SeqCst);
```

Comment:
I find it hard to reason about this. What does this context track? It's specific for compaction, so it should be all reads done during compaction, but what are those?

Reply:
All reads done during compaction. We use <number after compaction> - <number before compaction> to estimate the number of kbs read in the process.